include/irr/builtin/material_compiler/glsl/rasterization/impl.glsl

```diff
@@ -3,150 +3,237 @@
 
-void instr_eval_execute(in instr_t instr, in vec3 L)
+void instr_eval_execute(in instr_t instr, in MC_precomputed_t precomp, inout irr_glsl_LightSample s, inout irr_glsl_AnisotropicMicrofacetCache _microfacet, in bool skip)
 {
-	uint op = instr_getOpcode(instr);
-
-	//speculative execution
-	bsdf_data_t bsdf_data = fetchBSDFDataForInstr(instr);
-	params_t params = instr_getParameters(instr, bsdf_data);
-	float bxdf_eval_scalar_part;
-	uint ndf = instr_getNDF(instr);
-	float a = params_getAlpha(params);
-	float a2 = a*a;
-#ifndef ALL_ISOTROPIC_BXDFS
-	float ay = params_getAlphaV(params);
-	float ay2 = ay*ay;
-#endif
+	const uint op = instr_getOpcode(instr);
+	const bool is_bxdf = op_isBXDF(op);
+	const bool is_bsdf = !op_isBRDF(op); //note it actually tells if op is BSDF or BUMPMAP or SET_GEOM_NORMAL (divergence reasons)
+	const float cosFactor = irr_glsl_conditionalAbsOrMax(is_bsdf, s.NdotL, 0.0);
+	const bool positiveCosFactor = cosFactor > FLT_MIN;
+	const bool is_bxdf_or_combiner = op_isBXDForCoatOrBlend(op);
 
-	float cosFactor = op_isBSDF(op) ? abs(currBSDFParams.isotropic.NdotL):max(currBSDFParams.isotropic.NdotL,0.0);
+	uvec3 regs = instr_decodeRegisters(instr);
+	mat2x3 ior;
+	mat2x3 ior2;
+	params_t params;
+	irr_glsl_AnisotropicMicrofacetCache microfacet;
+	bsdf_data_t bsdf_data;
 
-	if (cosFactor>FLT_MIN && op_hasSpecular(op))
+	const bool run = !skip;
+
+	if (run && is_bxdf_or_combiner)
 	{
-#ifdef NDF_GGX
-#ifndef ONLY_ONE_NDF
-		if (ndf==NDF_GGX) {
-#endif
+		bsdf_data = fetchBSDFDataForInstr(instr);
+		ior = bsdf_data_decodeIoR(bsdf_data, op);
+		ior2 = matrixCompMult(ior, ior);
+		params = instr_getParameters(instr, bsdf_data);
+	}
 
-#ifdef ALL_ISOTROPIC_BXDFS
-			bxdf_eval_scalar_part = irr_glsl_ggx_height_correlated_cos_eval_DG(currBSDFParams.isotropic, currInteraction.isotropic, a2);
-#else
-			bxdf_eval_scalar_part = irr_glsl_ggx_height_correlated_aniso_cos_eval_DG(currBSDFParams, currInteraction, a, ay);
-#endif
+	bxdf_eval_t bxdf_eval = bxdf_eval_t(0.0);
 
+	if (run && is_bxdf && positiveCosFactor)
+	{
+		const float eta = colorToScalar(ior[0]);
+		const float rcp_eta = 1.0 / eta;
 
-#ifndef ONLY_ONE_NDF
-		} else
+		const float NdotV = irr_glsl_conditionalAbsOrMax(is_bsdf, currInteraction.isotropic.NdotV, 0.0);
+
+		bool is_valid = (NdotV > FLT_MIN);
+		bool refraction = false;
+#ifdef OP_DIELECTRIC
+		if (op == OP_DIELECTRIC && irr_glsl_isTransmissionPath(currInteraction.isotropic.NdotV, s.NdotL))
+		{
+			is_valid = irr_glsl_calcAnisotropicMicrofacetCache(microfacet, true, currInteraction.isotropic.V.dir, s.L, currInteraction.T, currInteraction.B, currInteraction.isotropic.N, s.NdotL, s.VdotL, eta, rcp_eta);
+			refraction = true;
+		}
+		else
+#endif
+		{
+			microfacet = _microfacet;
+		}
+
+#if defined(OP_DIELECTRIC) || defined(OP_CONDUCTOR)
+		is_valid = irr_glsl_isValidVNDFMicrofacet(microfacet.isotropic, is_bsdf, refraction, s.VdotL, eta, rcp_eta);
+#endif
```

Comment:
a function where you don't need to compute the PDF (such as rasterization's EVAL only), does not need this check (its redundant)